manual: minor editorial fixes in tutorials/parallelism.etex

gvrooyen opened the pull request:
1. Made several minor language fixes throughout the _Parallel Programming_ chapter.
2. Added external reference links to some parallel programming concepts (work stealing, Treiber stack) that may be unfamiliar to some readers.

gvrooyen committed:
manual: minor editorial fixes in tutorials/parallelism.etex

dra27 commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -2,23 +2,23 @@
 \label{c:parallelism}
 
-In this chapter, we shall look at the parallel programming facilities in OCaml.
+In this chapter we will look at the parallel programming facilities in OCaml.
```
Or just drop the word entirely? I don't mind the tense disagreement with the actual tutorial.
```suggestion
In this chapter we look at the parallel programming facilities in OCaml.
```

dra27 commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -123,7 +123,7 @@ Benchmark 1: ./fib_twice.exe 42
 
-As one can see that computing the nth Fibonacci number twice almost took the
+As one can see, computing the $n$th Fibonacci number twice takes almost the
 same time as computing it once thanks to parallelism.
```
Comma here??
```suggestion
same time as computing it once, thanks to parallelism.
```

dra27 commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -175,8 +175,9 @@ tutorial uses domainslib version 0.5.0.
-parallel implementation of the Fibonacci program is given below.
+\href{https://en.wikipedia.org/wiki/Work_stealing}{work-stealing queues} in
+order to efficiently share tasks with other domains. A parallel implementation
+of the Fibonacci program is given below.
```
Or, cf. change above:
```suggestion
of the Fibonacci program follows:
```
?

dra27 commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -223,9 +224,9 @@ For small inputs, the "fib_par" function simply calls the sequential Fibonacci
-For simplicity, we use "ocamlfind" to compile this program. It is recommended
-that the users use \href{https://github.com/ocaml/dune}{dune} to build their
-programs that utilise libraries installed through
+For simplicity, we use "ocamlfind" to compile this program. In general it is
```
```suggestion
For simplicity, we use "ocamlfind" to compile this program. In general, it is
```

dra27 commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -418,7 +419,7 @@ allocated. Having domain-local pools avoids synchronisation for most major heap
 algorithm that involves a few short stop-the-world pauses for each major cycle.
 
-Overall, the users should expect the garbage collector to scale well with
+Overall, users should expect the garbage collector to scale well with an
 increasing number of domains, with the latency remaining low. For more
```
Or (subjectively):
```suggestion
Overall, users should expect the garbage collector to scale well with as
the number of domains increases, with latency remaining low. For more
```

dra27 commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -464,7 +465,7 @@ Domains may perform blocking synchronisation with the help of
 modules. These modules are the same as those used to synchronise threads
 created by the threads library (chapter \ref{c:threads}). For clarity, in the
-rest of this chapter, we shall call the threads created by the threads library
+rest of this chapter, we will call the threads created by the threads library
```
Again, I'd drop it here:
```suggestion
rest of this chapter, we call the threads created by the threads library
```
or
```suggestion
rest of this chapter, we refer to the threads created by the threads library
```

dra27 commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -592,21 +593,21 @@ This program uses a shared reference cell protected by a mutex to communicate
 between the different systhreads running on two different domains. The
 systhread identifiers uniquely identify systhreads in the program. The initial
 domain gets the domain id and the thread id as 0. The newly spawned domain gets
```
Or:
```suggestion
domain gets the domain id 0 and thread id 0. The newly spawned domain gets
```
(and similarly in the next sentence?)?

dra27 commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -592,21 +593,21 @@ This program uses a shared reference cell protected by a mutex to communicate
-to manipulate global C state (e.g. initialise a function-local static value).
-This is no longer true in the presence of parallel execution with multiple
-domains.
+to manipulate the global C state (e.g. initialise a function-local static
+value). This is no longer true in the presence of parallel execution with
+multiple domains.
```
This change isn't necessary (this is the only change I actually object to)

gvrooyen replied:
Haha no dispute in matters of taste! 🙂 I'll revert to the original wording

dra27 commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -737,5 +738,5 @@ If the "compare_and_set" fails, then some other domain is also attempting to
 allowing competing domains to make progress before retrying the failed
-operation. This lock-free stack implementation is also known as Treiber
-stack.
+operation. This lock-free stack implementation is also known as the 
```
I'd personally put "a" rather than "the", but I'm actually wondering if the whole sentence is not quite there - is the point more that the lock-free stack _is_ a Treiber stack, rather than being known as one?

gvrooyen replied:
How about "this lock-free stack implementation is called a Treiber Stack"?

dra27 commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -39,24 +39,24 @@ thread. Each domain also has its own runtime state, which includes domain-local
 
-\emph{\textbf{It is recommended that the programs do not spawn more domains
+\emph{\textbf{It is recommended that programs do not spawn more domains
 than cores available}}.
```
```suggestion
than the number of available cores}}.
```
?

dra27 reviewed (commented):
Thank you for this - tweaking language of manuals is of course something that can be hugely subjective, but I like the changes here (with one exception, noted in the review). Some of the additional comments I make are just on reading the text (yet) again, and are very much suggestions.

I have worked (in person) with @kayceesrk for quite a few years, and some of the original text I "hear" as him! @kayceesrk - how do the changes look to you?

dra27 commented:
@Octachron had privately me asked about the s/shall/will - I think the original "shall" is valid (in fact, pedantically, I think it's more correct than "will" in the contexts in which it appears here), but to me it is more formal than intended and "will" does feel more natural.

gvrooyen committed:
Update manual/src/tutorials/parallelism.etex

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

gvrooyen committed:
Update manual/src/tutorials/parallelism.etex

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

gvrooyen committed:
Update manual/src/tutorials/parallelism.etex

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

gvrooyen committed:
Update manual/src/tutorials/parallelism.etex

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

gvrooyen committed:
Update manual/src/tutorials/parallelism.etex

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

gvrooyen committed:
Update manual/src/tutorials/parallelism.etex

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

gvrooyen committed:
Update manual/src/tutorials/parallelism.etex

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

gvrooyen committed:
Update manual/src/tutorials/parallelism.etex

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

kayceesrk commented:
> I have worked (in person) with @kayceesrk for quite a few years, and some of the original text I "hear" as him! @kayceesrk - how do the changes look to you?

I'm not a native English speaker and I do have my own style of writing :-) I don't have strong opinions on editorial changes. If others feel that this is an improvement, please feel free to go ahead with the changes.

kayceesrk commented:
I quickly went through the changes, and they do look like an improvement.

gvrooyen commented:
> @Octachron had privately me asked about the s/shall/will - I think the original "shall" is valid (in fact, pedantically, I think it's more correct than "will" in the contexts in which it appears here), but to me it is more formal than intended and "will" does feel more natural.

Yes, it's a more formal form, and "will" is more common nowadays. Here I was just guided by the style and tone of the rest of the OCaml documentation.

gvrooyen commented:
> I quickly went through the changes, and they do look like an improvement.

Thanks, @kayceesrk! Also let me just say that I thoroughly enjoyed reading this chapter – thank you for doing such a clear and interesting write-up 😁

gvrooyen committed:
Update parallelism.etex

Phrasing "domain id 0" on line 596.
Reverted "the global C state" to "global C state".

gvrooyen committed:
Update parallelism.etex

Simplified the wording around Treiber stack.

gvrooyen commented:
Okay, edits as suggested have been made. Thanks for all the feedback!

kayceesrk commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -593,17 +593,17 @@ This program uses a shared reference cell protected by a mutex to communicate
 systhread identifiers uniquely identify systhreads in the program. The initial
 domain gets the domain id 0 and thread id 0. The newly spawned domain gets
-the domain id as 1.
+the domain id 0.
```
Is this correct? I believe this should remain `1`.

gvrooyen replied:
Oh goodness, that's a typo! Thanks for catching it

dbuenzli commented:
Ah since we are here. I was reading this very chapter a few days ago. And got confused by:
```
Concurrency is overlapped execution of tasks
(section \ref{s:effects-concurrency}) whereas parallelism is simultaneous
execution of tasks
```
Isn't that *overlapped* supposed to be *interleaved* ?

gvrooyen committed:
Update parallelism.etex

Fixed type in newly spawned domain id

gvrooyen commented:
> Ah since we are here. I was reading this very chapter a few days ago. And got confused by:
> 
> ```
> Concurrency is overlapped execution of tasks
> (section \ref{s:effects-concurrency}) whereas parallelism is simultaneous
> execution of tasks
> ```
> 
> Isn't that _overlapped_ supposed to be _interleaved_ ?

I believe _interleaved_ is the better descriptor, but @kayceesrk is the domain expert here so is the best person to confirm.

kayceesrk commented:
Interleaved is indeed better.

gvrooyen committed:
Update parallelism.etex

Changed description of concurrency as "interleaved" rather than "overlapped".

gvrooyen commented:
I think that's everything resolved then, @dra27? Thank you again for everyone's input

dra27 reviewed (approved):
Lovely, thank you!

dra27 commented:
Thanks everybody - pushed to 5.3 as well, so it should be included in the manual release after 5.3.0 is done.
